src/Pate/Discovery.hs

```diff
@@ -309,58 +301,81 @@ concreteValueAddress ::
   MC.Value arch ids (MT.BVType (MC.ArchAddrWidth arch)) ->
   [PA.ConcreteAddress arch]
 concreteValueAddress = \case
-  MC.RelocatableValue _ addr -> [PA.ConcreteAddress addr]
+  MC.RelocatableValue _ addr -> [ PA.memAddrToAddr addr ]
   MC.BVValue w bv |
     Just WI.Refl <- WI.testEquality w (MM.memWidthNatRepr @(MC.ArchAddrWidth arch)) ->
-      [PA.ConcreteAddress (MM.absoluteAddr (MM.memWord (fromIntegral bv)))]
+      [ PA.memAddrToAddr (MM.absoluteAddr (MM.memWord (fromIntegral bv))) ]
   MC.AssignedValue (MC.Assignment _ rhs) -> case rhs of
     MC.EvalApp (MC.Mux _ _ b1 b2) -> concreteValueAddress b1 ++ concreteValueAddress b2
     _ -> []
   _ -> []
+  -- TODO ^ is this complete?
+
 
 concreteJumpTargets ::
-  forall bin arch ids m.
-  Monad m =>
+  forall sym bin arch ids.
+  HasCallStack =>
   PB.KnownBinary bin =>
   PA.ValidArch arch =>
+  PB.ConcreteBlock arch bin ->
   MD.ParsedBlock arch ids ->
-  m [BlockTarget arch bin]
-concreteJumpTargets pb = case MD.pblockTermStmt pb of
-  MD.ParsedCall st ret -> go (concreteNextIPs st) ret
-  MD.PLTStub st _ _ -> case MapF.lookup (MC.ip_reg @(MC.ArchReg arch)) st of
-    Just addr -> go (concreteValueAddress addr) Nothing
-    _ -> return $ []
-  MD.ParsedJump _ tgt -> do
-    blk <- mkConcreteBlock PB.BlockEntryJump tgt
-    return $ [ BlockTarget blk Nothing ]
-  MD.ParsedBranch _ _ t f -> do
-    blk_t <- mkConcreteBlock PB.BlockEntryJump t
-    blk_f <- mkConcreteBlock PB.BlockEntryJump f
-    return $ [ BlockTarget blk_t Nothing, BlockTarget blk_f Nothing ]
-  MD.ParsedLookupTable _jt st _ _ -> go (concreteNextIPs st) Nothing
+  EquivM sym arch [PB.BlockTarget arch bin]
+concreteJumpTargets from pb = case MD.pblockTermStmt pb of
+  MD.ParsedCall st ret ->
+    callTargets from (concreteNextIPs st) ret PB.BlockEntryPostFunction
+
+  MD.PLTStub st _ _ ->
+    case MapF.lookup (MC.ip_reg @(MC.ArchReg arch)) st of
```

Comment:
For `RegState`, you can use the `boundValue` lens (https://github.com/GaloisInc/macaw/blob/master/base/src/Data/Macaw/CFG/Core.hs#L589). `st ^. boundValue (ip_reg @arch)`.

This is conceptually total (our guarantees of this are well-intentioned and usually realized)

Reply:
Is there a particular reason PLTStub has a `MapF (ArchReg arch) (Value arch ids)` instead of a `RegState (ArchReg arch) (Value arch ids)`?

Reply:
Damn, I didn't realize that wasn't a `RegState`. It looks like the `PLTStub` version is actually a sparse map.

This case might just not make sense (and I'd be surprised if we are actually executing these stubs). Even if you found a PC value there, it would probably be to a computed value in a shared library. It is not the address returned to. Unfortunately, `PLTStub`s can't know where they are returning to.

During execution, the program issues a call to the PLT stub, putting the return address either on the stack or into a link register. The PLT stub then tail calls into a shared library, which will issue the return.

Reply:
I'm not sure what that means for this function.  Should we change this case?

Reply:
My hypothesis is that this case should just be a panic - I don't think we ever reach it. If we do reach it, we don't have a way to handle it properly. The way it should be is that, if we see a call to a shared library function through a PLT stub, we just make an "external call" frame using that `ExternalDomain` type. If we aren't doing that right now, that is a separate bug.

Reply:
OK, I'll make this an error and see what happens to the test suite.

Reply:
Well, changing this breaks several test cases in the ppc test suite. I don't really know what that means.

Reply:
Which tests? PPC has not been thoroughly tested yet, so that may not be a huge problem